// @devvit/public-api/context-actions/ContextActionBuilder.js
var _ContextActionsBuilder_actions;
import { __classPrivateFieldGet } from "tslib";
import { ContextActionDescription, ContextActionList, } from '@devvit/protos';
/**
 * Helper class to simplify the creation of a ContextActionDescription response
 */
export class ContextActionsBuilder {
    constructor() {
        _ContextActionsBuilder_actions.set(this, []);
    }
    build() {
        return ContextActionList.fromPartial({
            actions: __classPrivateFieldGet(this, _ContextActionsBuilder_actions, "f"),
        });
    }
    action(opts) {
        if (!opts.post && !opts.comment && !opts.subreddit) {
            throw new Error('Must specify at least one type');
        }
        __classPrivateFieldGet(this, _ContextActionsBuilder_actions, "f").push(ContextActionDescription.fromPartial({
            actionId: opts.actionId,
            name: opts.name,
            description: opts.description,
            contexts: {
                post: !!opts.post,
                comment: !!opts.comment,
                subreddit: !!opts.subreddit,
            },
            users: {
                moderator: !!opts.moderator,
                member: !!opts.member,
            },
            userInput: opts.userInput,
        }));
        return this;
    }
}
_ContextActionsBuilder_actions = new WeakMap();
